File: VerbumKit/Sources/Features/Account/ProfileFeature.swift
import Clients
import ComposableArchitecture
import Foundation
import Models

enum ProfileAppearance: String, CaseIterable { case automatic, light, dark }

extension SharedKey where Self == AppStorageKey<ProfileAppearance>.Default {
    static var profileAppearance: Self { Self[.appStorage("profileAppearance"), default: .automatic] }
}
extension SharedKey where Self == FileStorageKey<ReadingActivity>.Default {
    static var readingActivity: Self {
        Self[.fileStorage(.documentsDirectory.appending(component: "reading-activity.json")), default: ReadingActivity()]
    }
}

@Reducer
struct ProfileFeature {
    @ObservableState
    struct State: Equatable {
        @Shared(.readingActivity) var activity
        @Shared(.profileAppearance) var appearance
        @Shared(.lastRead) var lastRead
        var usage: UsageStatus?
        var usageFailed = false
        var annotations: [ReaderAnnotation] = []
        var loading = false
        var loadFailed = false
        var readingSettings = ReaderSettingsFeature.State()
        var notifications = DailyVerseFeature.State()
        var bookmarks: [ReaderAnnotation] { annotations.filter { $0.bookmarked == true } }
        var highlights: [ReaderAnnotation] { annotations.filter { $0.highlight != nil } }
        var notes: [ReaderAnnotation] { annotations.filter { !$0.note.isEmpty } }
    }
    enum Action: Equatable {
        case task, loaded([ReaderAnnotation]), loadFailed
        case usageLoaded(UsageStatus?), usageFailed
        case appearanceChanged(ProfileAppearance)
        case readingSettings(ReaderSettingsFeature.Action)
        case notifications(DailyVerseFeature.Action)
    }
    @Dependency(\.usageClient) var usageClient
    @Dependency(\.readerAnnotations) var annotations
    var body: some ReducerOf<Self> {
        Scope(state: \.readingSettings, action: \.readingSettings) { ReaderSettingsFeature() }
        Scope(state: \.notifications, action: \.notifications) { DailyVerseFeature() }
        Reduce { state, action in
            switch action {
            case .task:
                state.loading = true; state.loadFailed = false
                return .merge(.run { [usageClient] send in
                    do { await send(.usageLoaded(try await usageClient.status())) } catch { await send(.usageFailed) }
                }, .run { [annotations] send in
                    do { await send(.loaded(try await annotations.load())) }
                    catch is CancellationError {} catch { await send(.loadFailed) }
                }.cancellable(id: "profile-annotations", cancelInFlight: true))
            case .usageLoaded(let value): state.usage = value; state.usageFailed = false; return .none
            case .usageFailed: state.usageFailed = true; return .none
            case .loaded(let values): state.annotations = values; state.loading = false; return .none
            case .loadFailed: state.loading = false; state.loadFailed = true; return .none
            case .appearanceChanged(let appearance): state.$appearance.withLock { $0 = appearance }; return .none
            case .readingSettings, .notifications: return .none
            }
        }
    }
}

File: android/feature/scripture/src/main/kotlin/com/nexussoft/verbum/feature/scripture/ProfileFeature.kt
package com.nexussoft.verbum.feature.scripture

import com.nexussoft.verbum.clients.*
import com.nexussoft.verbum.common.arch.*
import com.nexussoft.verbum.models.*
import kotlinx.coroutines.CancellationException

object ProfileFeature {
    const val APPEARANCE_KEY = "profileAppearance"
    enum class Appearance(val copyKey: String) { AUTOMATIC("automatic"), LIGHT("light"), DARK("dark") }
    data class State(
        val usage: UsageStatus? = null,
        val usageFailed: Boolean = false,
        val activity: ReadingActivity = ReadingActivity(),
        val lastRead: PassageReference? = null,
        val annotations: List<ReaderAnnotation> = emptyList(),
        val loading: Boolean = false,
        val failed: Boolean = false,
        val appearance: Appearance = Appearance.AUTOMATIC,
        val settings: ReaderSettingsFeature.State = ReaderSettingsFeature.State(ReaderTextScale.STANDARD),
    ) {
        val bookmarks get() = annotations.filter { it.bookmarked }
        val highlights get() = annotations.filter { it.highlight != null }
        val notes get() = annotations.filter { it.note.isNotEmpty() }
    }
    sealed interface Action {
        data class UsageLoaded(val value: UsageStatus?): Action
        data object UsageFailed: Action
        data object Started: Action
        data class Loaded(val activity: ReadingActivity, val annotations: List<ReaderAnnotation>, val settings: ReaderSettingsFeature.State, val lastRead: PassageReference?): Action
        data object Failed: Action
        data class AppearanceChanged(val value: Appearance): Action
        data class Settings(val action: ReaderSettingsFeature.Action): Action
    }
    fun initial(preferences: PreferencesClient) = State(appearance = runCatching {
        Appearance.valueOf(preferences.string(APPEARANCE_KEY).orEmpty())
    }.getOrDefault(Appearance.AUTOMATIC))

    fun reducer(preferences: PreferencesClient, annotations: ReaderAnnotationsClient = PreferenceReaderAnnotationsClient(preferences), usageStatus: suspend () -> UsageStatus? = { null }): Reducer<State, Action> = combine(
        ReaderSettingsFeature.reducer(preferences).pullback(
            get = { it.settings }, set = { s, c -> s.copy(settings = c) },
            extractAction = { (it as? Action.Settings)?.action }, embedAction = { Action.Settings(it) },
        ),
        Reducer { state, action ->
            when(action) {
                Action.Started -> state.copy(loading = true, failed = false).with(runEffect(id = "profile-load", cancelInFlight = true) { send ->
                    try {
                        send(Action.Loaded(ReadingActivityClient(preferences).load(), annotations.load(), ReaderSettingsFeature.State(
                            ReaderTextScale.fromPreference(preferences.string(ReaderTextScale.PREFERENCE_KEY)),
                            runCatching { ReadingMode.valueOf(preferences.string(ChapterReaderFeature.MODE_KEY).orEmpty()) }.getOrDefault(ReadingMode.PAGES),
                            preferences.string(ChapterReaderFeature.FOCUS_KEY) == "true",
                        ), preferences.string(ChapterReaderFeature.LAST_READ_KEY)?.let(LastRead::decode)))
                    } catch (error: CancellationException) { throw error }
                    catch (_: Exception) { send(Action.Failed) }
                    try { send(Action.UsageLoaded(usageStatus())) }
                    catch(error: CancellationException) { throw error }
                    catch(_: Exception) { send(Action.UsageFailed) }
                })
                is Action.UsageLoaded -> state.copy(usage = action.value, usageFailed = false).only()
                Action.UsageFailed -> state.copy(usageFailed = true).only()
                is Action.Loaded -> state.copy(activity = action.activity, annotations = action.annotations, settings = action.settings, lastRead = action.lastRead, loading = false).only()
                Action.Failed -> state.copy(loading = false, failed = true).only()
                is Action.AppearanceChanged -> state.copy(appearance = action.value).with(runEffect { preferences.setString(APPEARANCE_KEY, action.value.name) })
                is Action.Settings -> state.only()
            }
        },
    )
}
